### DISCUSSION TITLE
NaN in the training and extremely high number of points sampled compared to num_domain, num_test

### DISCUSSION THREAD

**Discussion by dang-trinh156** [lululxvi/deepxde#1254] [Q&A] [answered]
Dr Mr. LuLu,

THE FIRST PROBLEM I ENCOUNTERED IS:
I am solving a complicated PDE - Modified Darcy equation for 2 phase Oil-Water: 
![image](https://user-images.githubusercontent.com/124559501/232515013-5cbad313-2400-458e-908d-d33a0c409c8b.png)
![image](https://user-images.githubusercontent.com/124559501/232515181-fc2db933-7e8b-4171-9da6-125932d1c58c.png)
where I predict Oil Pressure (Po) and Water Saturation(Sw), (u = [Po, Sw]). Moreover, Relative Permeability of oil and water phases (Kro,Krw) is a function of Sw; and μw, Bw,Bo,μo is a function of Oil Pressure. Here, I compute Kro,Krw, μw, Bw,Bo,μo by using tf.where to adjust specific elements with simple operations (multiply or divide or plus or minus Sw,Po with a real number, not exponential or sqrt of Sw,Po). Everything seems normal before training (IC,BC, dde.data.TimePDE, model.compile run without errors). But when I traing, I got this warning along with nan values in the 1000th epoch (maybe it has already be nan in the 2nd epoch). 
![image](https://user-images.githubusercontent.com/124559501/232518194-0538c9a2-b03f-4862-a5bc-d627f229e01b.png)

Here is the code:
```
import deepxde as dde
import matplotlib.pyplot as plt
import datetime
import time
import numpy as np
import tensorflow as tf

geom = dde.geometry.geometry_3d.Cuboid(xmin = [0.,0.,8325.], xmax =[10000.,10000.,8425.] )
timedomain = dde.geometry.TimeDomain(0,120)
geomtime = dde.geometry.GeometryXTime(geom, timedomain)

dx = 1000  #Length of a grid block in x-axis
dy = 1000  #Length of a grid block in y-axis
beta_c=1.127  #Unit conversion number, just a number
alpha_c = 5.614583 #Unit conversion number, just a number
phi = 0.3 #Porosity, it's just a real number

def PDE(x,u): #Reservoir simulation PDE calling function outside dde.grads
    dz = function(x[:,:3] # dz is computed based on location of the grid cell, this takes x,y,z as input and returns an array same dimension with derivative of Po
    kx = permxyz(x[:,:3]) # same as function computing dz
    Ax = dy*dz; #Area that flow goes through according to each cartaxis
    Vb = dx*dy*dz  #Volume of a grid
    qo,qw = rates(x[:,:3])  # rates is computed based on location of the grid cell, this takes x,y,z as input and returns an array same dimension with derivative of Po for each elements (qo, qw)
    Po, Sw= u[:, 0:1], u[:, 1:2]
    P_x = dde.grad.jacobian(Po, x, j=0)
 '''Here I compute each derivative of Kro,Bo,Bw... individually and then put them into a symbol derived function from the two PDEs above '''
    oil = symbolic derived form of above PDE
    water = symbolic derived form of above PDE

    return [oil, water]
ic = dde.icbc.IC(
    geomtime,
    lambda X: 6000,
    lambda _, on_initial: on_initial)
bc = dde.icbc.NeumannBC(geomtime, lambda X: 0, lambda _, on_boundary: on_boundary)
geom = dde.geometry.geometry_3d.Cuboid(xmin = [0.,0.,8325.], xmax =[10000.,10000.,8425.] )
timedomain = dde.geometry.TimeDomain(0,120)
geomtime = dde.geometry.GeometryXTime(geom, timedomain)
data = dde.data.pde.TimePDE(
    geomtime,
    flowContinuity,
    [
        bc,
        ic,
    ],
    num_domain=300,
    num_boundary=40,
    num_initial=40,
    num_test=200,
)
net = dde.nn.FNN([4] + 3 * [50] + [2], "tanh", "Glorot normal")
model = dde.Model(data, net)
model.compile("adam", lr=1.0e-3)
model.train(iterations=3000)
```

THE SECOND PROBLEM IS:
when I define the `data = dde.data.pde.TimePDE(
    geomtime,
    flowContinuity,
    [
        bc,
        ic,
    ],
    num_domain=300,
    num_boundary=40,
    num_initial=40,
    num_test=200,
)` from the above code, I only need num_domain,num_boundary, num_initial, num_test like I specified above but got this warning:
![image](https://user-images.githubusercontent.com/124559501/232522744-0c2f4ccd-588a-45d1-acaf-5af1e52224d0.png)
This is exponentially higher than the number of points that I want to specify.

Thank you for your answer!

**Comment by lululxvi** [comment]
Why do you only use 200 points for testing?

**Reply by dang-trinh156** [reply]
As in my problem, there's only 300 interior points (300 centroids, when i use PointCloud for input arrays), so I use those number. But , even if i set it to 1, the number is still incredibly large

**Comment by lululxvi** [comment, marked as answer]
For the NaN problem, your problem setup doesn't have a good scale, see FAQ Q: I failed to train the network or get the right solution, e.g., large training loss, unbalanced losses.

**Reply by dang-trinh156** [reply]
Thank you for the answer, but what about the large number of points sampled compared to defined ones?

**Reply by lululxvi** [reply]
I cannot understand your question of "the large number of points sampled compared to defined ones"

**Reply by dang-trinh156** [reply]
My apologies for unclarity, 
Here I define num_domain = 300,num_boundary = 40, num_initial = 40, num_test = 200, but I dont understand why I got warning : 129100 points required ,  but 165675 points sampled, Warning: 200 points required , 165675 points sampled

**Reply by lululxvi** [reply]
The test points are in a **uniform** grid. There is no way to generate uniform grid points by only using 300 points, so more points are used automatically based on the calculation.

**Comment by Elyasin97** [comment]
@dang-trinh156 Hi I'm currently working on a similar project, and I believe your code could provide valuable insights and guidance. If you're comfortable with it, I was wondering if you could share some snippets of your final code.
thank you

**Reply by dang-trinh156** [reply]
Unfortunately, as I encountered various errors that have high possibility coming from simulation basics, I've been exploring the numerical simulation field recently.  Therefore, I haven't finished the code yet.
But I believe we can exchange our experiences for this subject.
